src/docker/image.rs

```diff
@@ -0,0 +1,227 @@
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use crate::TargetTriple;
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+pub struct Image {
+    pub name: String,
+    // The toolchain triple the image is built for
+    pub toolchain_triple: ImageArchitecture,
+}
+
+impl std::fmt::Display for Image {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.name)
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
+pub struct PossibleImage {
+    pub name: String,
+    // The toolchain triple the image is built for
+    pub toolchain: Option<ImageArchitecture>,
+}
+
+impl PossibleImage {
+    pub(crate) fn to_definite_with(&self, architecture: &ImageArchitecture) -> Image {
+        if let Some(toolchain_triple) = &self.toolchain {
+            Image {
+                name: self.name.clone(),
+                toolchain_triple: toolchain_triple.clone(),
+            }
+        } else {
+            Image {
+                name: self.name.clone(),
+                toolchain_triple: architecture.clone(),
+            }
+        }
+    }
+}
+
+impl FromStr for PossibleImage {
+    type Err = std::convert::Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(PossibleImage {
+            name: s.to_string(),
+            toolchain: None,
+        })
+    }
+}
+
+impl std::fmt::Display for PossibleImage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.name)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(try_from = "&str")]
+pub struct ImageArchitecture {
+    pub architecture: Architecture,
+    pub os: Os,
+    pub variant: Option<String>,
```

Comment:
Can you document what variant is or an example? Everything else is pretty self-explanatory, and I think I get it (having multiple different images for the same architecture, so they're tagged differently), but it might be good when reading through the codebase in a few months.

Reply:
absolutely, it's basically what docker uses (which is how go does it)

ill leave a comment